File: Cpp/Test/Tapenade/ba.c
#include "ba.h"
#include <stdlib.h>
#include <math.h>
#include <float.h>

double sqsum(int n, double *x)
{
  int i;
  double res;
  res = 0;
  for (i = 0; i < n; i++)
    res = res + x[i] * x[i];
  return res;
}

void cross(double *a, double *b, double *out)
{
  out[0] = a[1] * b[2] - a[2] * b[1];
  out[1] = a[2] * b[0] - a[0] * b[2];
  out[2] = a[0] * b[1] - a[1] * b[0];
}

void rodrigues_rotate_point(double *rot, double *pt, double *rotatedPt)
{
  int i;
  double sqtheta, theta, costheta, sintheta, theta_inverse,
    w[3], cross_[3], tmp;

  sqtheta = sqsum(3, rot);
  if (sqtheta != 0.)
  {
    theta = sqrt(sqtheta);
    costheta = cos(theta);
    sintheta = sin(theta);
    theta_inverse = 1.0 / theta;

    for (i = 0; i < 3; i++)
      w[i] = rot[i] * theta_inverse;

    cross(w, pt, cross_);

    tmp = (w[0] * pt[0] + w[1] * pt[1] + w[2] * pt[2]) *
      (1. - costheta);

    for (i = 0; i < 3; i++)
      rotatedPt[i] = pt[i] * costheta + cross_[i] * sintheta + w[i] * tmp;
  }else
  {
    cross(rot, pt, cross_);
    
    for (i = 0; i < 3; i++)
      rotatedPt[i] = pt[i] + cross_[i];
  }
}

void radial_distort(double *rad_params, double *proj)
{
  double rsq, L;
  rsq = sqsum(2, proj);
  L = 1 + rad_params[0] * rsq + rad_params[1] * rsq * rsq;
  proj[0] = proj[0] * L;
  proj[1] = proj[1] * L;
}

void project(double *cam, double *X, double *proj)
{
  int i;
  double *C;
  double Xo[3], Xcam[3];
  C = &cam[BA_C_IDX];

  for (i = 0; i < 3; i++)
    Xo[i] = X[i] - C[i];

  rodrigues_rotate_point(&cam[BA_ROT_IDX], Xo, Xcam);

  proj[0] = Xcam[0] / Xcam[2];
  proj[1] = Xcam[1] / Xcam[2];

  radial_distort(&cam[BA_RAD_IDX], proj);

  for (i = 0; i < 2; i++)
    proj[i] = proj[i] * cam[BA_F_IDX] + cam[BA_X0_IDX + i];
}

void computeReprojError(
  double *cam,
  double *X, 
  double *w, 
  double feat_x, 
  double feat_y,
  double *err)
{
  double proj[2];
  project(cam, X, proj);

  err[0] = (*w)*(proj[0] - feat_x);
  err[1] = (*w)*(proj[1] - feat_y);

  // This term is here so that tapenade correctly 
  // recognizes inputs to be the inputs
  err[0] = err[0] + (cam[0] - cam[0]) + (X[0] - X[0]);
}

void computeZachWeightError(double *w, double *err)
{
  *err = 1 - (*w)*(*w);
}

void ba_objective(int n, int m, int p, 
  double *cams, 
  double *X,
  double *w, 
  int *obs, 
  double *feats,
  double *reproj_err, 
  double *w_err)
{
  int i, camIdx, ptIdx;

  for (i = 0; i < p; i++)
  {
    camIdx = obs[i * 2 + 0];
    ptIdx = obs[i * 2 + 1];
    computeReprojError(&cams[camIdx * BA_NCAMPARAMS], &X[ptIdx * 3],
      &w[i], feats[i * 2 + 0], feats[i * 2 + 1], &reproj_err[2 * i]);
  }

  for (i = 0; i < p; i++)
  {
    computeZachWeightError(&w[i], &w_err[i]);
  }

  // This term is here so that tapenade correctly 
  // recognizes inputs to be the inputs
  //reproj_err[0] = reproj_err[0] + ((cams[0] - cams[0]) +
  //  (X[0] - X[0]) + (w[0] - w[0]));
}

File: Fsharp/DiffSharp/ba.fs
module ba

open System
open System.Diagnostics
open System.IO

#if MODE_AD
open DiffSharp.AD
#else
open DiffSharp.AD.Specialized.Reverse1
#endif

let N_CAM_PARAMS = 11
let ROT_IDX = 0
let CENTER_IDX = 3
let FOCAL_IDX = 6
let X0_IDX = 7
let RAD_IDX = 9

////// Objective //////

let inline sqnorm (x:_[]) =
    x |> Array.map (fun x -> x*x) |> Array.sum

let inline dot_prod (x:_[]) (y:_[]) =
    Array.map2 (*) x y |> Array.sum

let inline sub_vec (x:_[]) (y:_[]) =
    Array.map2 (-) x y

let inline add_vec (x:_[]) (y:_[]) =
    Array.map2 (+) x y

let inline add_vec3 (x:_[]) (y:_[]) (z:_[]) =
    Array.map3 (fun a b c -> a+b+c) x y z

let inline mult_by_scalar (x:_[]) y =
    Array.map (fun a -> a*y) x

let inline cross (a:_[]) (b:_[]) =
    [|a.[1]*b.[2] - a.[2]*b.[1]; a.[2]*b.[0] - a.[0]*b.[2]; a.[0]*b.[1] - a.[1]*b.[0];|]

let rodrigues_rotate_point (rot:_[]) (X:_[]) =
    let sqtheta = sqnorm rot
    if sqtheta <> 0. then
        let theta = sqrt sqtheta
        let costheta = cos theta
        let sintheta = sin theta
        let theta_inv = 1. / theta

        let w = mult_by_scalar rot theta_inv
        let w_cross_X = cross w X    
        let tmp = (dot_prod w X) * (1. - costheta)

        add_vec3 (mult_by_scalar X costheta) (mult_by_scalar w_cross_X sintheta) (mult_by_scalar w tmp)
    else
        add_vec X (cross rot X)

let radial_distort (rad_params:_[]) (proj:_[]) =
    let rsq = sqnorm proj
    let L = 1. + rad_params.[0] * rsq + rad_params.[1] * rsq * rsq
    mult_by_scalar proj L

let project (cam:_[]) (X:_[]) =
    let Xcam = rodrigues_rotate_point cam.[ROT_IDX..(ROT_IDX+2)] (sub_vec X cam.[CENTER_IDX..(CENTER_IDX+2)])
    let distorted = radial_distort cam.[RAD_IDX..(RAD_IDX+1)] (mult_by_scalar Xcam.[0..1] (1./Xcam.[2]))
    add_vec cam.[X0_IDX..(X0_IDX+1)] (mult_by_scalar distorted cam.[FOCAL_IDX])

let compute_reproj_err (cam:_[]) (X:_[]) w (feat:_[]) =
    mult_by_scalar (sub_vec (project cam X) feat) w

let compute_zach_weight_error w =
    1. - w*w

let ba_objective (cams:_[][]) (X:_[][]) (w:_[]) (obs:int[][]) (feat:float[][]) =
    let n = cams.Length
    let p = w.Length
    let reproj_err = 
        [|for i = 0 to p-1 do yield (compute_reproj_err cams.[obs.[i].[0]] X.[obs.[i].[1]] w.[i] feat.[i])|]
    let w_err = Array.map compute_zach_weight_error w 
    reproj_err, w_err

///// Derivative extras /////
let rodrigues_rotate_point_ (rot:D[]) (X:D[]) =
    let sqtheta = sqnorm rot
    let zero = rot.[0] - rot.[0] // simple (D 0.) did not work
    if sqtheta <> zero then
        let theta = sqrt sqtheta
        let costheta = cos theta
        let sintheta = sin theta
        let theta_inv = 1. / theta

        let w = mult_by_scalar rot theta_inv
        let w_cross_X = cross w X    
        let tmp = (dot_prod w X) * (1. - costheta)

        add_vec3 (mult_by_scalar X costheta) (mult_by_scalar w_cross_X sintheta) (mult_by_scalar w tmp)
    else
        add_vec X (cross rot X)

let radial_distort_ (rad_params:D[]) (proj:D[]) =
    let rsq = sqnorm proj
    let L = 1. + rad_params.[0] * rsq + rad_params.[1] * rsq * rsq
    mult_by_scalar proj L

let project_ (cam:D[]) (X:D[]) =
    let Xcam = rodrigues_rotate_point_ cam.[ROT_IDX..(ROT_IDX+2)] (sub_vec X cam.[CENTER_IDX..(CENTER_IDX+2)])
    let distorted = radial_distort_ cam.[RAD_IDX..(RAD_IDX+1)] (mult_by_scalar Xcam.[0..1] (1./Xcam.[2]))
    add_vec cam.[X0_IDX..(X0_IDX+1)] (mult_by_scalar distorted cam.[FOCAL_IDX])

let compute_reproj_err_ (cam:D[]) (X:D[]) (w:D) (feat:float[]) =
    mult_by_scalar (sub_vec (project_ cam X) feat) w
    
let compute_reproj_err_wrapper (parameters:_[]) (feat:float[]) =
    let X_off = N_CAM_PARAMS
    let w_off = X_off + 3
    compute_reproj_err_ parameters.[..(X_off-1)] parameters.[X_off..(X_off+2)] parameters.[w_off] feat

let vectorize (cam:_[]) (X:_[]) (w:_) = 
    Array.append (Array.append cam X) [|w|]

let compute_zach_weight_error_ (w:D) =
    1. - w*w
    
////// UTILS ////

// TODO
let create_sparse_J n m p obs (reproj_err_d:_[,][]) (w_err_d:_[]) =
    let nrows = 2 * p + p;
    let ncols = N_CAM_PARAMS*n + 3 * m + p;

//    let n_new_cols = N_CAM_PARAMS + 3 + 1
//    let rows1 = [|for i=0 to 2*p-1 do yield i*n_new_cols|];
//    let w_off = n_new_cols*2*p
//    let rows2 = [|for i=0 to p-1 do yield i + w_off+1|];
//    let rows = Array.append (Array.append [|0|] rows1) rows2
    
    (reproj_err_d, w_err_d)
    //(nrows,ncols,rows,cols,vals)

////// IO //////

let read_ba_instance (fn:string) =
    let read_in_elements (fn:string) =
        let string_lines = File.ReadLines(fn)
        let separators = [|' '|]
        [| for line in string_lines do yield line.Split separators |] 
            |> Array.map (Array.filter (fun x -> x.Length > 0))

    let parse_double (arr:string[]) = Array.map Double.Parse arr

    let data = read_in_elements fn

    let n = Int32.Parse data.[0].[0]
    let m = Int32.Parse data.[0].[1]
    let p = Int32.Parse data.[0].[2]
    
    let mutable offset = 1
    let one_cam = parse_double data.[offset]
    let cams = [|for i=1 to n do yield one_cam|]
    offset <- offset + 1

    let one_X = parse_double data.[offset]
    let X = [|for i=1 to m do yield one_X|]
    offset <- offset + 1

    let one_w = Double.Parse data.[offset].[0]
    let w = [|for i=1 to p do yield one_w|]
    offset <- offset + 1

    let one_feat = parse_double data.[offset]
    let feat = [|for i=1 to p do yield one_feat|]

    let mutable camIdx = 0
    let mutable ptIdx = 0
    let obs = [|for i=1 to p do 
                yield [|camIdx; ptIdx|]
                camIdx <- (camIdx+1) % n
                ptIdx <- (ptIdx+1) % m
              |]

    cams, X, w, obs, feat